### PR TITLE
[NameLookup] ASTOOScope ontology

### PR DESCRIPTION
**Pull request by davidungar**
Initial cut at an ontology for scopes. 
Disabled by default, but you can pass `-enable-astscope-lookup` to turn it on, or set an environment variable: `export SWIFT_TEST_OPTIONS=-enable-astscope-lookup` to tell `lit` to pass that option. There's also `-Xfrontend -compare-to-astscope-lookup` which will fail an assertion if the new lookup varies from the old.

Parser-based lookup is still used for now, even with the new lookup.

### CODE REVIEW

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@slavapestov @DougGregor Here's a PR with the scopes. This code definitely should be refactored before it lands. In particular, I'm not sure that the use of class templates worked that well because I could not get the method instantiations to work as I wanted them to. Also, it's possible that multiple inheritance could clean it up a lot. It *does* manage to find the right lookup decls and baseDC's for all of the .swift files in the test directory. I'm starting to work on the IndexOfFirstOuterResults: there's one test where it differs. Next, will be getting the dependency info to match, then seeing how many tests pass or fail with the new lookup code & also performance work.
Thanks!

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
At this point, the code seems to get the right answer for IndexOfFirstOuterResult, too. Next stop: the dependency info.

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by gottesmm** [comment]
This is a huge patch, are you going to split this up?

**Review by gottesmm — COMMENTED** [review]
Just some drive by comments. This is a really huge patch, it is important that it is split up. There is also a bunch of formatting (i.e. indentation) errors that would be caught by git-clang-format.

**Comment by davidungar** [comment]
@swift-ci please clean smoke test os x platform

File: include/swift/AST/ASTScope.h [hunk 1]
```diff
@@ -1,4 +1,5 @@
-//===--- ASTScope.h - Swift AST Scope ---------------------------*- C++ -*-===//
+//===--- ASTScopeImpl.h - Swift AST Object-Oriented Scope -----------*- C++
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: Tx. Fixed

File: include/swift/AST/ASTScope.h [hunk 2]
```diff
@@ -10,117 +11,75 @@
 //
 //===----------------------------------------------------------------------===//
 //
-// This file defines the ASTScope class and related functionality, which
+// This file defines the ASTScopeImpl class ontology, which
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: Tx for the pointer. Fixed.

File: include/swift/AST/ASTScope.h [hunk 3]
```diff
@@ -10,117 +11,75 @@
...
+//   a child's source range is a subset (proper or improper) of its parent's,
+//   children are ordered by source range, and do not overlap,
+//   all the names visible within a parent are visible within the child, unless
+//   the nesting is illegal. For instance, a protocol nested inside of a class
+//   does not get to see the symbols in the class or its ancestors.
+//
 //===----------------------------------------------------------------------===//
-#ifndef SWIFT_AST_AST_SCOPE_H
-#define SWIFT_AST_AST_SCOPE_H
+#ifndef SWIFT_AST_AST_OO_SCOPE_H
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: Tx. Slipped by when I did a global rename.

File: include/swift/AST/ASTScope.h [hunk 4]
```diff
@@ -10,117 +11,75 @@
...
+class DeclName;
+
+namespace ast_scope {
+class ASTScopeImpl;
+class GTXScope;
+class IterableTypeScope;
+class TypeAliasScope;
+class DeferredNodes;
+
+#pragma mark the root ASTScopeImpl class
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: I like the way they help my eyes parse the file. Also Xcode supports them. Is there a substitute?

File: include/swift/AST/ASTScope.h [hunk 5]
```diff
@@ -10,117 +11,75 @@
...
+class TrailingWhereClause;
+class ParameterList;
+class PatternBindingEntry;
+class SpecializeAttr;
+class GenericContext;
+class DeclName;
+
+namespace ast_scope {
+class ASTScopeImpl;
+class GTXScope;
```
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: The Doxygen comment for GTXScope says GenericType or Extension Scope. I know it's a new abbreviation to the code base -- I've seen NTD used a bunch -- and I generally agree with spelling things out. If I spell it out, we get:
`GenericTypeOrExtensionScope`
`GenericTypeOrExtensionWherePortion`
`GenericTypeOrExtensionWholePortion.`
Seems to be getting a bit long, but I'm sympathetic to the idea of spelling it out.

[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: I'll make sure that each class with GTX in its name includes a comment with an explanation of the abbreviation.

**Comment by davidungar** [comment]
@gottesmm Thanks for taking a look! As far as splitting it up, the one piece that I can figure out to split off is the extra source location info for `EditorPlaceHolder`. Other than that, I don't see any way to split it up. It is code that is off-by-default, so I don't expect any destabilization. It is also code that is new, so I don't foresee any conflicts.

File: include/swift/AST/Expr.h [hunk 6]
```diff
@@ -4666,16 +4664,19 @@ class UnresolvedPatternExpr : public Expr {
 class EditorPlaceholderExpr : public Expr {
   Identifier Placeholder;
   SourceLoc Loc;
+  SourceLoc TrailingAngleBracketLoc;
```
Comment (gottesmm) [inline include/swift/AST/Expr.h]: Here is an example of something that could be ripped out into a separate patch. This could be sliced off.

File: include/swift/AST/NameLookup.h [hunk 7]
```diff
@@ -36,52 +37,91 @@ namespace swift {
...
+  class PatternBindingInitializer;
+  class DefaultArgumentInitializer;
+  class NominalTypeDecl;
+
+  namespace ast_scope {
+  class ASTSourceFileScope;
+  class ASTScopeImpl;
+  } // namespace ast_scope
+
+  /// LookupResultEntry - One result of unqualified lookup.
```
Comment (gottesmm) [inline include/swift/AST/NameLookup.h]: Why are you indenting this out, that just increases the diff without need.

File: include/swift/AST/NameLookup.h [hunk 8]
```diff
@@ -110,13 +150,13 @@ class UnqualifiedLookup {
   /// is used to determine which declarations in that body are visible.
   UnqualifiedLookup(DeclName Name, DeclContext *DC, LazyResolver *TypeResolver,
                     SourceLoc Loc = SourceLoc(), Options options = Options());
-
-  SmallVector<LookupResultEntry, 4> Results;
+  using ResultsVector = SmallVector<LookupResultEntry, 4>;
```
Comment (gottesmm) [inline include/swift/AST/NameLookup.h]: Are you using this type outside of the class? If not, why create this? If you are using it outside, it is better to use SmallVectorImpl.

File: include/swift/AST/NameLookup.h [hunk 9]
```diff
@@ -110,13 +150,13 @@ class UnqualifiedLookup {
...
+  ResultsVector Results;
   /// The index of the first result that isn't from the innermost scope
```
Comment (gottesmm) [inline include/swift/AST/NameLookup.h]: Please put a newline line here.

File: include/swift/AST/NameLookup.h [hunk 10]
```diff
@@ -536,8 +575,91 @@ class FindLocalVal : public StmtVisitor<FindLocalVal> {
...
+  virtual bool consume(ArrayRef<ValueDecl *> values, DeclVisibilityKind vis,
+                       NullablePtr<DeclContext> baseDC = nullptr) = 0;
+
+  /// Eventually this functionality should move into ASTScopeLookup
+  virtual std::pair<bool, Optional<bool>>
+  lookupInSelfType(NullablePtr<DeclContext> selfDC, DeclContext *const scopeDC,
+                   NominalTypeDecl *const nominal,
+                   Optional<bool> isCascadingUse) = 0;
+
+  virtual void stopForDebuggingIfTargetLookup() = 0;
```
Comment (gottesmm) [inline include/swift/AST/NameLookup.h]: Is this going to be called in the code always? If this is only for debugging the compiler, it should be hidden behind an `#ifdef`.

File: include/swift/AST/NameLookup.h [hunk 11]
```diff
@@ -536,8 +575,91 @@ class FindLocalVal : public StmtVisitor<FindLocalVal> {
...
+public:
+  ASTScope(ast_scope::ASTSourceFileScope *sfs) : impl(sfs) {}
+  static ASTScope *createScopeTreeFor(SourceFile *);
+  static Optional<bool>
+  unqualifiedLookup(SourceFile *, DeclName, SourceLoc,
+                    const DeclContext *startingContext,
+                    Optional<bool> isCascadingUse,
+                    namelookup::AbstractASTScopeDeclConsumer &);
+
+  void dump() const;
```
Comment (gottesmm) [inline include/swift/AST/NameLookup.h]: Can you make sure that the dump methods are only in assert builds? That is what we generally do.

File: include/swift/Basic/STLExtras.h [hunk 12]
```diff
@@ -833,6 +833,13 @@ class RelationalOperationsBase {
     return !(left == right);
   }
 };
+  
+/// Cast a pointer to \c U  to a pointer to a supertype \c T.
+/// Example:  Wobulator *w = up_cast<Wobulator>(coloredWobulator)
+/// Useful with ?: where each arm is a different subtype.
+/// If \c U is not a subtype of \c T, the compiler will complain.
+template <typename T, typename U>
+T *up_cast(U *ptr) { return ptr; }
```
Comment (gottesmm) [inline include/swift/Basic/STLExtras.h]: We follow camelCase for functions.

File: lib/AST/ASTScopeCreation.cpp [hunk 13]
```diff
@@ -0,0 +1,1061 @@
...
+/// As scopes are created, pass down the \c ASTNodes in enclosing \c Decls
+/// that really need to be in subscopes.
+///
+/// Because parsing and lookup (typechecking) are interleaved for some files,
+/// we need to be able to resume scope tree creation. \c DeferredNodes also
+/// helps with this by remembering the \c lastAdopter.
+
+class DeferredNodes final {
+  /// The nodes that must be passed down to deeper levels in the scope tree.
+  std::deque<ASTNode> nodes;
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Do you need to use an std::deque? I have heard of it having sub-par performance IIRC.

File: lib/AST/ASTScopeCreation.cpp [hunk 14]
```diff
@@ -0,0 +1,1061 @@
...
+  /// lookup, create scopes for any \c Decls beyond this number.
+  int numberOfDeclsAlreadyProcessed = 0;
+
+  /// The last scope to "adopt" deferred nodes.
+  /// When adding \c Decls to a scope tree that have been created since the tree
+  /// was originally built, add them as children of this scope.
+  NullablePtr<ASTScopeImpl> lastAdopter;
+
+  /// Information about how this object was created for debugging.
+  const char *const file;
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: If this is for debugging, it should be ifdefed out without asserts. Also, you probably want to use a StringRef.

File: lib/AST/ASTScopeCreation.cpp [hunk 15]
```diff
@@ -0,0 +1,1061 @@
...
+  const int line;
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Please do not use a const value type. That is something in general in LLVM style we avoid.

File: lib/AST/ASTScopeCreation.cpp [hunk 16]
```diff
@@ -0,0 +1,1061 @@
...
+public:
+
+#pragma mark ASTNodes that do not create scopes
+
+  // Even ignored Decls and Stmts must extend the source range of a scope:
+  // E.g. a braceStmt with some definitions that ends in a statement that
+  // accesses such a definition must resolve as being IN the scope.
+  // clang-format off
+                          
+  void visitImportDecl(                       ImportDecl *d, ASTScopeImpl *p, DeferredNodes&) { p->widenSourceRangeForIgnoredASTNode(d); }
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Can you use a `#define` here to avoid reproducing the code? E.g.:

```
#ifdef SCOPE_CREATOR_IGNORED
#error "Already created"
#endif
#define SCOPE_CREATOR_IGNORED(NAME) \
void visit##Name(Name *d, ASTScopeImpl *p, DeferredNodes &) { \
  p->widenSourceRangeForIgnoredASTNode(d); \
}
SCOPE_CREATOR_IGNORED(ImportDecl)
SCOPE_CREATOR_IGNORED(EnumCaseDecl)
...
#undef SCOPE_CREATOR_IGNORED
```

File: lib/AST/ASTScopeCreation.cpp [hunk 17]
```diff
@@ -0,0 +1,1061 @@
...
+ void visitPoundAssertStmt( PoundAssertStmt *s, ASTScopeImpl *p, DeferredNodes&) { p->widenSourceRangeForIgnoredASTNode(s); }
+
+    // clang-format on
+                          
+
+#pragma mark simple creation ignoring deferred nodes
+// clang-format off
+  void visitSubscriptDecl(SubscriptDecl *e, ASTScopeImpl *p, DeferredNodes &) { p->createSubtree<SubscriptDeclScope>(e); }
+
+  void visitIfStmt(                   IfStmt *e, ASTScopeImpl *p, DeferredNodes &) { p->createSubtree<     IfStmtScope>(e); }
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Same here. It would be much easier to read and wouldn't have special formatting if you use a `#define` like I suggested above.

File: lib/AST/ASTScopeCreation.cpp [hunk 18]
```diff
@@ -0,0 +1,1061 @@
...
+  }
+}
+
+template <typename StmtExpr>
+void ASTScopeImpl::dispatchAndCreateIfNeeded(StmtExpr *se,
+                                             DeferredNodes &deferred) {
+  if (se)
+    ScopeCreator().visit(se, this, deferred);
+}
+void ASTScopeImpl::dispatchAndCreateIfNeeded(Decl *d, DeferredNodes &deferred) {
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Can you make sure to put spaces in between things like this?

File: lib/AST/ASTScopeCreation.cpp [hunk 19]
```diff
@@ -0,0 +1,1061 @@
...
+}
+
+void ConditionalClauseScope::expandMe(DeferredNodes &deferred) {
+  // If this is a boolean conditional not in a guard continuation, add a
+  // child for the expression.
+  createSubtreeForCondition();
+  // If there's another conditional clause, add it as the child.
+  if (index + 1 < getContainingStatement()->getCond().size())
+    createSubtreeForNextConditionalClause(deferred);
+  else {
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Everywhere else you did if () { } else { }. Can you be consistent with that here.

File: lib/AST/ASTScopeCreation.cpp [hunk 20]
```diff
@@ -0,0 +1,1061 @@
...
+
+void WholeClosureScope::expandMe(DeferredNodes &) {
+  if (auto *cl = captureList.getPtrOrNull())
+    createSubtree<CaptureListScope>(cl);
+  if (closureExpr->getInLoc().isValid())
+    createSubtree<ClosureParametersScope>(captureList, closureExpr);
+  createSubtree<ClosureBodyScope>(captureList, closureExpr);
+}
+
+
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Another inconsistency around newlines.

File: lib/AST/ASTScopeCreation.cpp [hunk 21]
```diff
@@ -0,0 +1,1061 @@
...
+  assert(expr);
+
+  /// AST walker that finds top-level closures in an expression.
+  class ClosureFinder : public ASTWalker {
+    llvm::function_ref<void(NullablePtr<CaptureListExpr>, ClosureExpr *)>
+        foundClosure;
+
+  public:
+    ClosureFinder(
+        llvm::function_ref<void(NullablePtr<CaptureListExpr>, ClosureExpr *)>
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Just a friendly FYI, I injected function_ref into the swift namespace a bit ago. You can drop the llvm::.

File: lib/AST/ASTScopeCreation.cpp [hunk 22]
```diff
@@ -0,0 +1,1061 @@
...
+            foundClosure)
+        : foundClosure(foundClosure) {}
+
+    std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+      if (auto closure = dyn_cast<ClosureExpr>(E)) {
```
Comment (gottesmm) [inline lib/AST/ASTScopeCreation.cpp]: Can you put a * on the pointer auto to make it clear it is a pointer.

File: lib/AST/ASTScopeCreation.cpp (hunk 13 above)
Comment (gottesmm) [reply lib/AST/ASTScopeCreation.cpp]: Also, another thing couldn't the DeferredNodes be committed as a separate commit? It is separable, no?

**Review by gottesmm — COMMENTED** [review]
More feedback, some questions and some suggestions on stuff that I think could be split out of the patch.

**Comment by davidungar** [comment]
@swift-ci please clean smoke test os x platform

File: include/swift/AST/Expr.h (hunk 6 above)
Comment (davidungar) [reply include/swift/AST/Expr.h]: Yes, agreed! Our comments crossed in the ether.

File: include/swift/AST/NameLookup.h (hunk 7 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: Probably a git-clang-format thing--I'll fix.

File: include/swift/AST/NameLookup.h (hunk 8 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: Yes,` UnqualifedLookupFactory` uses it, and no, the Impl type won't do. But is there a better way?

File: include/swift/AST/NameLookup.h (hunk 9 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: Yup, good catch, thanks!

File: lib/AST/ASTScopeCreation.cpp (hunk 15 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Why avoid it? I think it helps make the code more readable.

File: lib/AST/ASTScopeCreation.cpp (hunk 13 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: What would be the benefit in committing it separately? It is only used by the new scope code.

File: include/swift/Basic/STLExtras.h (hunk 12 above)
Comment (davidungar) [reply include/swift/Basic/STLExtras.h]: But `up_cast` is consistent with `dyn_cast`, etc.

File: lib/AST/ASTScopeCreation.cpp (hunk 22 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Good catch, thanks!

File: lib/AST/ASTScopeCreation.cpp (hunk 21 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Cool! I'll fix all my uses.

File: include/swift/AST/NameLookup.h (hunk 10 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: Good point! Will do.

File: include/swift/AST/NameLookup.h (hunk 11 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: I understand the desire to slim down the compiler and sometimes `dump` is benind `ifndef NDBUG`. But also I see places where `void() dump` seems to appear all the time, and I worry about losing that debugging facility in release builds. In the code of Doug's that I started with, he used: `LLVM_ATTRIBUTE_DEPRECATED` and I'll do that for my `dump`.

File: lib/AST/ASTScopeCreation.cpp (hunk 13 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: There was some reason for `deque`, but I'll look at it again.

File: lib/AST/ASTScopeCreation.cpp (hunk 15 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Also prevents bugs. What's the reason to avoid?

File: include/swift/AST/NameLookup.h (hunk 8 above)
Comment (gottesmm) [reply include/swift/AST/NameLookup.h]: Why wont it do. If you are passing it to someone, you can use SmallVectorImpl< LookupResultEntry> &

File: lib/AST/ASTScopeCreation.cpp (hunk 13 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Good, thanks! A vector will do.

Comment (gottesmm) [reply lib/AST/ASTScopeCreation.cpp]: The benefit is by splitting up the patch into smaller pieces, it becomes easier to review. A big blob like this has historically resulted in us missing some pretty bad bugs.

Comment (gottesmm) [reply lib/AST/ASTScopeCreation.cpp]: Keep in mind, I am /not/ talking about committing it separately, I am talking about as a separate commit in the same PR. So you can have multiple commits in the PR that make it easy to review the code separately.

File: include/swift/AST/NameLookup.h (hunk 8 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: In `unqualifiedLookup.cpp`:
```
  if (compareToASTScopes && useASTScopesForExperimentalLookupIfEnabled()) {
    ResultsVector results;
    size_t indexOfFirstOuterResult = 0;
    UnqualifiedLookupFactory scopeLookup(Name, DC, TypeResolver, Loc, options,
                                         results, indexOfFirstOuterResult);
```
The code needs to create a results vector, so it can make another factory for comparison purposes. Publishing the type seemed like the clearest way. I suppose the code could use a `decltype` instead but it's not clearer.

File: lib/AST/ASTScopeCreation.cpp (hunk 13 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Oh, no I did not understand that. I misread `splitting up the patch into smaller pieces` to mean `splitting up the patch into smaller pull requests`. I can split it up into multiple commits, but if each one is non-functional till the final one, how does that help? If you are reviewing a comment that introduces DeferredNodes, but that type isn't used anywhere, how would it be reviewed?

On the other hand, I do think it's worth while to split off the trailing-angle-bracket-location-for-EditorPlaceholders as a separate PR, so that folks like @rintaro can look at that in isolation.
It homes in on his expertise and the added functionality is self-evident.

File: lib/AST/ASTScopeCreation.cpp (hunk 14 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Yes, agree about the ifdeffing. What's the benefit of a `StringRef` over a `const char*` here? These are truly compile-time constants, being the filename of the .cpp file in the compiler. A `const char*` is cheaper here, not that it matters much. 
I'll add some comments to make it clear that `file` is the compiler file.

File: lib/AST/ASTScopeCreation.cpp (hunk 16 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Yes, I had wondered about how folks here feel about defines? Makes the code shorter but harder to grep.

Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: OK, added defines (not pushed yet)

File: lib/AST/ASTScopeCreation.cpp (hunk 17 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Yup, done!

File: lib/AST/ASTScopeCreation.cpp (hunk 18 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Yes, a space there is good, thanks. Other places, I'm not so sure:
```
ASTScopeImpl *
ExtensionScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
  return parent->createSubtree2D<ExtensionScope, GTXWherePortion>(decl);
}
ASTScopeImpl *
NominalTypeScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
  return parent->createSubtree2D<NominalTypeScope, GTXWherePortion>(decl);
}
ASTScopeImpl *
TypeAliasScope::createTrailingWhereClauseScope(ASTScopeImpl *parent) {
  return parent->createSubtree2D<TypeAliasScope, GTXWherePortion>(decl);
}
```
Here, I could use a `define`, but it seems not quite an improvement for such a small number. So, I go with tight repetition.

File: lib/AST/ASTScopeCreation.cpp (hunk 19 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Yes, I see where I used braces but should not have. Thanks. I'm following a guideline from @jordan, which I quite like:
If the statement in the `if` (etc) is visually short (typ. one line) then braces just add clutter.
But if the statement wraps, then braces are better.

Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: I'll go through the code and eliminate bogus braces.

File: lib/AST/ASTScopeCreation.cpp (hunk 20 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Yup, fixed that one. But there will be places where I think that an extra newline can increase readability, just not there! Thanks.

File: lib/AST/ASTScopeCreation.cpp (hunk 19 above)
Comment (gottesmm) [reply lib/AST/ASTScopeCreation.cpp]: All I want is for you to be consistent about it.

**Comment by davidungar** [comment]
@gottesmm Thanks again! At this point, I have pushed all the changes I made based on your suggestions, in case you want to have another look.

File: lib/AST/ASTScopeCreation.cpp (hunk 19 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: Quite right and fair!

File: lib/AST/ASTScopeCreation.cpp (hunk 13 above)
Comment (davidungar) [reply lib/AST/ASTScopeCreation.cpp]: I agree that when each commit makes functional sense, it's easier to spot bugs. In this case, I'm not so sure because if DeferredNodes does the wrong thing, there's no way to know without reading the very code I would be splitting it off from.

However, I do think it might well serve, since DeferredNodes is declared in a .cpp file, to move the method bodies inside of the class definition.

@slavapestov ? What do you think? You're going to be reviewing for functionality, so it's important to accomadate you. @DougGregor, too. (@DougGregor `DeferredNodes` is my stab at more readable what-you-called "continuations", without the laziness constraint.

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

File: include/swift/AST/ASTScope.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/AST/ASTScope.h]: Please spell it out. How often are you going to read vs write this code (also since you use Xcode, I imagine the IDE will spell it out for you via autocomplete). For dinosaurs like me who do not use a new IDE, not having it spelled out makes life difficult.

I need to look at the other two ones. There may be a better name.

File: include/swift/AST/ASTScope.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (gottesmm) [reply include/swift/AST/ASTScope.h]: actually looking at some other parts of the frontend, if no one else has issues with it, then I am cool with it (looks like @DougGregor is using it a bunch).

File: include/swift/AST/ASTScope.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: OK--spelled it out.

File: include/swift/AST/ASTScope.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidungar) [reply include/swift/AST/ASTScope.h]: Great!
I *could* switch to // MARK if that's better

File: include/swift/AST/NameLookup.h (hunk 7 above)
Comment (davidungar) [reply include/swift/AST/NameLookup.h]: I think it's fixed. Let me know if not. I have tried to git-format it all but may have missed something.

**Comment by davidungar** [comment]
@DougGregor @slavapestov I have redone things according to my new pattern understanding. I think it's ready for your review(s) at this point. I've been banging on the code pretty hard, so you'll likely catch a bunch of things I missed. Thanks.

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test os x platform

**Comment by davidungar** [comment]
@swift-ci please smoke test

**Comment by davidungar** [comment]
@swift-ci please smoke test Linux platform

**Comment by davidungar** [comment]
Merged per offline conversation with @slavapestov

**Comment by compnerd** [comment]
This breaks on Windows! https://compnerd.visualstudio.com/windows-swift/_build/results?buildId=3205
